feat(batch-exports): Support concurrent uploads to S3

rossgray opened the pull request:
> [!IMPORTANT]
> 👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

## Problem

When there is a lot of data to export, S3 batch exports can often take a long time, even hitting the start to close timeout of the Temporal activity.

## Changes

Upload parts of the S3 multi part upload concurrently.

To support this we needed to rework the way we currently consume record batches. Rather than using a temporary file to write record batches, we keep these in memory.

This means we're using a Producer -> Transformer -> Consumer architecture:

- **Producer** -> reads Arrow data from our internal S3 stage and generates pyarrow RecordBatches
- **Transformer** -> transforms the RecordBatches to chunks of raw data (bytes), depending on the configured file format, compression, etc
- **Consumer** -> takes the raw data and writes it to the destination. In the case of S3, it uploads chunks of data to S3 concurrently, managing the initialization and completion of the multipart upload

**Note: this only applies to workflows using the new internal S3 stage, which makes it easier to decouple the reading of data from ClickHouse with the writing of data to S3.**

### TODO

- [x] Add more tests
- [x] Add more file types to the Transformer (can probably be done in a follow up PR)
- [x] Ensure S3 consumer has feature parity with existing one (eg handling retries)
- [x] Local testing

## Did you write or update any docs for this change?

<!-- Engineers are responsible for doing the first pass at documenting their features and/or code.  -->

- [ ] I've [added or updated the docs](https://posthog.com/handbook/engineering/writing-docs)
- [ ] I've reached out for help from the docs team
- [x] No docs needed for this change (can update them to explain internal workings once this functionality has been rolled out to all teams)

## How did you test this code?

- Added new tests
- Still need to do some local testing

rossgray committed:
WIP

rossgray committed:
WIP

rossgray committed:
Get scrappy version working

rossgray committed:
Cleanup

rossgray committed:
Revert changes to test

rossgray committed:
Support more features

rossgray committed:
fix bug not respecting exclude/include events

rossgray committed:
return records_completed

rossgray committed:
upload manifest file

rossgray committed:
WIP

tomasfarias commented on posthog/temporal/batch_exports/transformer.py:
```diff
@@ -0,0 +1,233 @@
+        """Write records to a temporary file as Parquet."""
+
+        self.parquet_writer.write_batch(record_batch.select(self.parquet_writer.schema.names))
+        # Get current buffer content
+        self._parquet_buffer.seek(0)
+        data = self._parquet_buffer.read()
+
+        # Reset buffer
+        self._parquet_buffer.seek(0)
+        self._parquet_buffer.truncate()
```
This feels wasteful: We are writing to a buffer, then copying the buffer data, and then deleting the buffer. Why not just write the data once, return that, and start a new buffer?

tomasfarias replied:
Saves the copy.

rossgray committed:
Fix handling of file splitting

tomasfarias commented on posthog/temporal/batch_exports/s3_batch_export.py:
```diff
@@ -970,23 +971,408 @@ async def insert_into_s3_activity_from_stage(inputs: S3InsertInputs) -> RecordsC
+        if not self.upload_id:
+            await self._initialize_multipart_upload()
+
+        # Extract part data
+        part_data = bytes(self.current_buffer[: self.part_size])
+        self.current_buffer = self.current_buffer[self.part_size :]
```
Either way, we need a limit. Similar to how we have `BATCH_EXPORT_S3_RECORD_BATCH_QUEUE_MAX_SIZE_BYTES` for the producer queue.

rossgray replied:
I'm a bit wary to pass indexes to a buffer around, since we will have multiple concurrent async tasks all reading from the same buffer, and would have to be very careful about how we reset the buffer.

The original implementation (suggested by Claude) had this `memory_semaphore` which I believe is intended to limit the memory consumption like you mention. I thought we could remove this and just limit memory via the concurrent uploads. However, as you say, I now realise we could create a big backlog of pending tasks that are all waiting to be uploaded, all of which would be consuming memory. So I'll add this back in

rossgray committed:
Update tests

tomasfarias commented on posthog/temporal/batch_exports/s3_batch_export.py:
```diff
@@ -970,23 +971,408 @@ async def insert_into_s3_activity_from_stage(inputs: S3InsertInputs) -> RecordsC
+
+    async def _cleanup_completed_uploads(self):
+        """Remove completed uploads from tracking"""
+        completed = []
+        for part_number, task in self.pending_uploads.items():
+            if task.done():
+                completed.append(part_number)
```
This could be a callback on the task.

tomasfarias commented on posthog/temporal/batch_exports/s3_batch_export.py:
```diff
@@ -970,23 +971,402 @@ async def insert_into_s3_activity_from_stage(inputs: S3InsertInputs) -> RecordsC
+            # self.memory_semaphore.release()
+            # Explicitly delete data to free memory immediately
+            del data
+            gc.collect()
```
I am not entirely convinced whether this is the right tradeoff. Sure, the GC will run faster, but it means we'll run it a lot more (once per upload part). So, we are going to be GC blocked a lot more.

rossgray committed:
Update tests

rossgray committed:
Make Parquet transformer more memory efficient

rossgray committed:
Cleanup running of consumer

rossgray committed:
Improve async task handling

rossgray committed:
Reuse s3 client

tomasfarias committed:
refactor(spmc): Reduce calls to record batch slice

Slicing a record batch is a zero copy operation, so no data is moved
around. However, we still pay some cycles for each call: In my
testing, execution time grew linearly with the size of a record batch.

This means we can optimize the `slice_record_batch` function if we
reduce the number of calls to `pyarrow.RecordBatch.slice`. The current
loop it's quite hot as it is taking 1 record out of the slice at a
time. Instead, we can take multiple rows at a time, by estimating how
many bytes we are still missing from the target.

rossgray committed:
add memory semaphore

rossgray committed:
Improve logging

rossgray committed:
Add S3 retries

rossgray commented on posthog/temporal/batch_exports/transformer.py:
```diff
@@ -0,0 +1,229 @@
+
+        self._brotli_compressor = None
+
+    def transform_batch(self, batch: pa.RecordBatch) -> typing.Generator[bytes, None, None]:
+        """Transform a single batch and yield compressed bytes"""
+        column_names = batch.column_names
+        if not self.include_inserted_at:
+            column_names.pop(column_names.index("_inserted_at"))
+
```
I will leave this as it's been working well until now, and is the same logic as used in `temporary_file.py`

rossgray commented on posthog/temporal/batch_exports/spmc.py:
```diff
@@ -1009,6 +1015,17 @@ def slice_record_batch(
+    if slice.nbytes <= max_record_batch_size_bytes or slice.num_rows <= min_records_per_batch:
+        return 0
+
+    avg_bytes_per_row = slice.nbytes / slice.num_rows
+    bytes_diff = slice.nbytes - max_record_batch_size_bytes
+    return max(math.floor(bytes_diff / avg_bytes_per_row), 1)
```
the check above should handle this:
```python
    if slice.nbytes <= max_record_batch_size_bytes or slice.num_rows <= min_records_per_batch:
        return 0
```

rossgray committed:
Fix type errors

rossgray committed:
Update tests

rossgray committed:
Remove TODO

rossgray committed:
fix test

tomasfarias commented on posthog/temporal/batch_exports/transformer.py:
```diff
@@ -0,0 +1,230 @@
+        """Write records to a temporary file as Parquet."""
+
+        self.parquet_writer.write_batch(record_batch.select(self.parquet_writer.schema.names))
+        # Get current buffer content without copying
+        data = self._parquet_buffer.getvalue()
```
This doesn't make sense: The comment says that this is getting "current buffer content without copying", but in the next two lines we cumplete delete everything from the buffer.

How, exactly, are we getting the data from the buffer without copying it? Data must live somewhere, but if it is still in the buffer then we delete it just below, and if it's not in the buffer, then we copied it somewhere?

rossgray replied:
yes, you're right, it does actually copy the data. I've removed this misleading comment. I'm not sure if there is a better way to handle this

rossgray committed:
update test

rossgray committed:
remove comments

tomasfarias commented on posthog/temporal/batch_exports/transformer.py:
```diff
@@ -0,0 +1,230 @@
+        yield data
+
+
+# TODO: Implement other transformers
```
Do we have any in mind?

rossgray replied:
yeah, Redshift too

tomasfarias commented on posthog/temporal/batch_exports/s3_batch_export.py:
```diff
@@ -970,23 +982,456 @@ async def insert_into_s3_activity_from_stage(inputs: S3InsertInputs) -> RecordsC
+        data_interval_start: dt.datetime | str | None,
+        data_interval_end: dt.datetime | str,
+        s3_inputs: S3InsertInputs,
+        part_size: int = 50 * 1024 * 1024,  # 50MB parts
+        max_concurrent_uploads: int = 5,
+        max_memory_buffer: int = 250 * 1024 * 1024,  # 250MB total buffer
```
Shouldn't `max_memory_buffer` be `part_size * max_concurrent_uploads`?

rossgray replied:
have removed this now

tomasfarias commented on posthog/temporal/batch_exports/s3_batch_export.py:
```diff
@@ -970,23 +982,456 @@ async def insert_into_s3_activity_from_stage(inputs: S3InsertInputs) -> RecordsC
+        part_number = self.part_counter
+        self.part_counter += 1
+
+        # Acquire memory semaphore (blocks if too many uploads in flight)
+        await self.memory_semaphore.acquire()
```
Isn't
```
async with self.upload_semaphore:
```
in `_upload_part_with_cleanup`, already doing this?

tomasfarias replied:
Yeah, so I would keep this one then. Since it will already limit the number of new tasks created, right?

tomasfarias replied:
You could even put the `release` as a callback of a task.

tomasfarias commented on posthog/temporal/batch_exports/s3_batch_export.py:
```diff
@@ -970,23 +982,456 @@ async def insert_into_s3_activity_from_stage(inputs: S3InsertInputs) -> RecordsC
+            # Explicitly delete data to free memory immediately
+            del data
+
+    def _get_current_key(self) -> str:
+        """Generate the key for the current file"""
+        key_prefix = get_s3_key_prefix(self.s3_inputs)
+
+        try:
+            file_extension = FILE_FORMAT_EXTENSIONS[self.s3_inputs.file_format]
+        except KeyError:
+            raise UnsupportedFileFormatError(self.s3_inputs.file_format, "S3")
+
+        base_file_name = f"{self.s3_inputs.data_interval_start}-{self.s3_inputs.data_interval_end}"
+        # to maintain backwards compatibility with the old file naming scheme
+        if self.s3_inputs.max_file_size_mb is not None:
+            base_file_name = f"{base_file_name}-{self.current_file_index}"
+        if self.s3_inputs.compression is not None:
+            file_name = base_file_name + f".{file_extension}.{COMPRESSION_EXTENSIONS[self.s3_inputs.compression]}"
+        else:
+            file_name = base_file_name + f".{file_extension}"
+
+        key = posixpath.join(key_prefix, file_name)
+
+        if posixpath.isabs(key):
+            # Keys are relative to root dir, so this would add an extra "/"
+            key = posixpath.relpath(key, "/")
+
+        return key
+
```
I think this the same as the `get_s3_key` function defined in line 143, why not use that?

tomasfarias replied:
Unless I am missing something, it looks the same

rossgray replied:
have switched to using the existing version

tomasfarias commented on posthog/temporal/batch_exports/s3_batch_export.py:
```diff
@@ -970,23 +982,456 @@ async def insert_into_s3_activity_from_stage(inputs: S3InsertInputs) -> RecordsC
+        try:
+            # Upload any remaining data in buffer
+            if len(self.current_buffer) > 0:
+                # for now, let's ignore this so our tests use multi part uploads too
+                # if not self.upload_id:
+                #     # Small file - direct upload
+                #     await self._single_file_upload_current()
+                # else:
+                # Upload final part
```
Do we need this uncommented for production? Could we have a 
```
if not settings.TEST and not settings.DEBUG:
...
```
so that it works both locally and in production?

rossgray replied:
could do, although then I suppose you're not testing the code that runs in production

rossgray committed:
use a single semaphore

rossgray committed:
Clean up

rossgray committed:
Add logging

rossgray committed:
More cleanup

rossgray committed:
Revert change to existing code
